### PR TITLE
Proper support for Snowflake TIMESTAMP_NTZ 

### LINKED ISSUE #1369: Unable to compare snowflake TIMESTAMP_NTZ type to Datafusion Timestamp type
**Describe the bug**

There is `Cannot infer common argument type for comparison operation Struct ...` error when I run queries comparing Snowflake TIMESTAMP_NTZ type value ([default type for Snowflake Timestamp and Date Time](https://docs.snowflake.com/en/sql-reference/data-types-datetime)) to Datafusion timestamp, for example 

>select "COLUMN_TIMESTAMP" from SF_TYPES WHERE "COLUMN_TIMESTAMP" < now();

Error
>Status { code: Internal, message: "type_coercion\ncaused by\nError during planning: Cannot infer common argument type for comparison operation Struct([Field { name: \"epoch\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {\"byteLength\": \"16\", \"finalType\": \"T\", \"logicalType\": \"TIMESTAMP_NTZ\", \"precision\": \"0\", \"charLength\": \"0\", \"scale\": \"9\"} }, Field { name: \"fraction\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {\"precision\": \"0\", \"scale\": \"9\", \"byteLength\": \"16\", \"finalType\": \"T\", \"charLength\": \"0\", \"logicalType\": \"TIMESTAMP_NTZ\"} }]) < Timestamp(Nanosecond, Some(\"+00:00\"))", metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Mon, 13 May 2024 16:06:16 GMT", "content-length": "0"} }, source: None }

**To Reproduce**

1. Create Snowflake table SF_TYPES with COLUMN_TIMESTAMP of type TIMESTAMP type which is an alias for TIMESTAMP_NTZ
2. Add Snowflake data connector for Spice
```
- from: snowflake:DEMO.PUBLIC.SF_TYPES
  name: sf_types
  params: 
  ...
```
4. Open Spice SQL and run query `select "COLUMN_TIMESTAMP" from SF_TYPES WHERE "COLUMN_TIMESTAMP" < now();` 
5. See error

>Status { code: Internal, message: "type_coercion\ncaused by\nError during planning: Cannot infer common argument type for comparison operation Struct([Field { name: \"epoch\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {\"byteLength\": \"16\", \"finalType\": \"T\", \"logicalType\": \"TIMESTAMP_NTZ\", \"precision\": \"0\", \"charLength\": \"0\", \"scale\": \"9\"} }, Field { name: \"fraction\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {\"precision\": \"0\", \"scale\": \"9\", \"byteLength\": \"16\", \"finalType\": \"T\", \"charLength\": \"0\", \"logicalType\": \"TIMESTAMP_NTZ\"} }]) < Timestamp(Nanosecond, Some(\"+00:00\"))", metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Mon, 13 May 2024 16:06:16 GMT", "content-length": "0"} }, source: None }

**Expected behavior**
I can run logical queries against Snowflake Timestamp and DateTime data types.

**Additional context**

Apache Spark has [TIMESTAMP_NTZ type as well](https://community.databricks.com/t5/technical-blog/introducing-timestamp-ntz-in-apache-spark-the-timestamp-without/ba-p/50586), support of this type across Snowflake and Spark/Databricks should be consistent.

Snowflake TIMESTAMP_NTZ is a struct
```
Struct([Field { name: "epoch", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {"byteLength": "16", "finalType": "T", "logicalType": "TIMESTAMP_NTZ", "precision": "0", "charLength": "0", "scale": "9"} }, Field { name: "fraction", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {"precision": "0", "scale": "9", "byteLength": "16", "finalType": "T", "charLength": "0", "logicalType": "TIMESTAMP_NTZ"} }]) 
```

Datafusion timestamp

``` 
describe select now()
+---------------+-------------------------------------------------------------------------------+
| plan_type     | plan                                                                          |
+---------------+-------------------------------------------------------------------------------+
| logical_plan  | Projection: TimestampNanosecond(1715621328263656000, Some("+00:00")) AS now() |
|               |   EmptyRelation                                                               |
| physical_plan | ProjectionExec: expr=[1715621328263656000 as now()]                           |
|               |   PlaceholderRowExec                                                          |
|               |                                                                               |
+---------------+-------------------------------------------------------------------------------+

```

### PR DESCRIPTION
**Pull request by sgrebnov**
Fixes [Unable to compare snowflake TIMESTAMP_NTZ type to Datafusion Timestamp type](https://github.com/spiceai/spiceai/issues/1369)

Snowflake TIMESTAMP_NTZ is received as struct. PRs implements logic to convert it to standard `DataType::Timestamp(TimeUnit::Millisecond, None)` type

### Snowflake

image

### Spice (with fix)

```
sql> SELECT "COLUMN_DATETIME", "COLUMN_DATE", "COLUMN_TIMESTAMP" FROM SF_TYPES
+---------------------+-------------+-------------------------+
| COLUMN_DATETIME     | COLUMN_DATE | COLUMN_TIMESTAMP        |
+---------------------+-------------+-------------------------+
| 2023-10-01T12:45:30 | 2024-05-02  | 2024-05-02T10:55:01.739 |
|                     |             |                         |
+---------------------+-------------+-------------------------+
```
```
sql> describe SF_TYPES
+------------------+------------------------------+-------------+
| column_name      | data_type                    | is_nullable |
+------------------+------------------------------+-------------+
| COLUMN_DATETIME  | Timestamp(Millisecond, None) | YES         |
| COLUMN_TIMESTAMP | Timestamp(Millisecond, None) | YES         |
+------------------+------------------------------+-------------+

```

### CODE REVIEW

File: crates/db_connection_pool/src/dbconnection/snowflakeconn.rs [hunk 1]
```diff
@@ -145,3 +169,69 @@ impl<'a> AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConne
...
+        }
+        fields.push(Arc::clone(field));
+        columns.push(Arc::clone(column));
+    }
+
+    let schema = Arc::new(Schema::new(fields));
+    RecordBatch::try_new(schema, columns).context(FailedToCreateRecordBatchSnafu)
+}
+
+fn cast_sf_timestamp_ntz_to_arrow_timestamp(column: &ArrayRef) -> Result<ArrayRef, Error> {
```
Comment (phillipleblanc) [inline crates/db_connection_pool/src/dbconnection/snowflakeconn.rs:203]: Let's add some tests for this function

Comment (sgrebnov) [reply crates/db_connection_pool/src/dbconnection/snowflakeconn.rs:203]: Added unit tests for `cast_sf_timestamp_ntz_to_arrow_timestamp` 
https://github.com/spiceai/spiceai/pull/1677/commits/b9effe6f0b4766a45fab2ed2500f83ca88a4ffaf

File: crates/db_connection_pool/src/dbconnection/snowflakeconn.rs [hunk 2]
```diff
@@ -145,3 +169,144 @@ impl<'a> AsyncDbConnection<Arc<SnowflakeApi>, &'a (dyn Sync)> for SnowflakeConne
...
+        ArrayBuilder, ArrayRef, Int32Builder, Int64Builder, StructBuilder,
+        TimestampMillisecondArray,
+    };
+    use arrow::datatypes::{DataType, Field};
+    use std::sync::Arc;
+
+    #[test]
+    fn test_cast_sf_timestamp_ntz_to_arrow_timestamp() {
+        let timestamp_ntz_array = create_timestamp_ntz_array(
+            vec![Some(1_696_164_330), None, Some(1_714_647_301)],
```
Comment (digadeesh) [inline crates/db_connection_pool/src/dbconnection/snowflakeconn.rs:258]: I'd test all of the permutations here.  (Some, Some)(none, none), (Some, None), (None, Some)

Comment (sgrebnov) [reply crates/db_connection_pool/src/dbconnection/snowflakeconn.rs:258]: Added test for invalid input 👍
https://github.com/spiceai/spiceai/pull/1677/files#diff-e0b90357e2dacc69cb483074b1b4a82a0d58da4d111f3f0179589a5053071b41R276

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**